src/ripple/app/misc/NegativeUNLVote.cpp

```diff
@@ -0,0 +1,354 @@
+    }
+
+    // have enough ledger ancestors, build the score table
+    for (auto const& k : unl)
+    {
+        scoreTable[k] = 0;
+    }
+    auto idx = numAncestors - 1;
+    for (int i = 0; i < 256; ++i)
+    {
+        for (auto const& v :
+             validations.getTrustedForLedger(ledgerAncestors[idx--]))
+        {
+            if (scoreTable.find(v->getNodeID()) != scoreTable.end())
+                ++scoreTable[v->getNodeID()];
+        }
+    }
+
+    unsigned int myValidationCount = 0;
+    if (scoreTable.find(myId_) != scoreTable.end())
+        myValidationCount = scoreTable[myId_];
+    if (myValidationCount < nUnlMinLocalValsToVote)
+    {
+        JLOG(j_.debug()) << "N-UNL: ledger " << seq << ". I only issued "
+                         << myValidationCount
+                         << " validations in last 256 ledgers."
+                         << " My reliability measurement could be wrong.";
+        return false;
+    }
+    else if (
+        myValidationCount > nUnlMinLocalValsToVote && myValidationCount <= 256)
+    {
+        return true;
+    }
+    else
+    {
+        // cannot happen because validations.getTrustedForLedger does not
+        // return multiple validations of the same ledger from a validator.
+        JLOG(j_.error()) << "N-UNL: ledger " << seq << ". I issued "
+                         << myValidationCount
+                         << " validations in last 256 ledgers. Too many!";
+        return false;
+    }
+}
+
+void
+NegativeUNLVote::findAllCandidates(
+    hash_set<NodeID> const& unl,
+    hash_set<NodeID> const& nUnl,
+    hash_map<NodeID, unsigned int> const& scoreTable,
+    std::vector<NodeID>& toDisableCandidates,
+    std::vector<NodeID>& toReEnableCandidates)
+{
+    /*
+     * -- Compute if need to find more validators to disable, by checking if
+     *    canAdd = sizeof nUnl < maxNegativeListed.
+     *
+     * -- Find toDisableCandidates: check if
+     *    (1) canAdd,
+     *    (2) has less than nUnlLowWaterMark validations,
+     *    (3) is not in nUnl, and
+     *    (4) is not a new validator.
+     *
+     * -- Find toReEnableCandidates: check if
+     *    (1) is in nUnl and has more than nUnlHighWaterMark validations
+     *    (2) if did not find any by (1), try to find candidates:
+     *        (a) is in nUnl and (b) is not in unl.
+     */
+    auto maxNegativeListed = (std::size_t)std::ceil(unl.size() * nUnlMaxListed);
+    std::size_t negativeListed = 0;
+    for (auto const& n : unl)
+    {
+        if (nUnl.find(n) != nUnl.end())
+            ++negativeListed;
+    }
+    bool canAdd = negativeListed < maxNegativeListed;
+    JLOG(j_.trace()) << "N-UNL: my nodeId " << myId_ << " lowWaterMark "
+                     << nUnlLowWaterMark << " highWaterMark "
+                     << nUnlHighWaterMark << " canAdd " << canAdd
+                     << " negativeListed " << negativeListed
+                     << " maxNegativeListed " << maxNegativeListed;
+
```

Comment:
I might change the above to a lamba that initialized `canAdd` and moves the rest into the lambda. Something like:
```
    auto const canAdd = [&]() -> bool {
        std::size_t negativeListed = 0;
        for (auto const& n : unl)
        {
            if (negUnl.find(n) != negUnl.end())
                ++negativeListed;
        }
        bool const result = negativeListed < maxNegativeListed;
        JLOG(j_.trace()) << "N-UNL: my nodeId " << myId_ << " lowWaterMark "
                         << nUnlLowWaterMark << " highWaterMark "
                         << nUnlHighWaterMark << " canAdd " << result
                         << " negativeListed " << negativeListed
                         << " maxNegativeListed " << maxNegativeListed;
        return result;
    }();
```

Reply:
I like this. 

Allow me a philosophical comment: I've come to dislike code like this:
```
        std::size_t negativeListed = 0;
        for (auto const& n : unl)
        {
            if (negUnl.find(n) != negUnl.end())
                ++negativeListed;
        }
```

While simple enough, I think that using the "shared vernacular" of `std::count_if` is much better, since it communicates the intent of the code precisely, with no loss in performance:

```
auto negativeListed = std::count_if(unl.begin(), unl.end(),
    [&negUnl](auto const& entry) {
        return negUnl.find(entry) != negUnl.end();
    });
```

The one downside is that it returns a signed type `std::unordered_map<...>::difference_type` which is a _**signed**_ type. The reasons for this escape me, but hey...
 
My comment isn't meant to be actionable. Just thinking out loud.

Reply:
Thanks, used the version in Scott's comment.